File: src/main/java/RentalCalculator.java
import io.vavr.Function1;
import io.vavr.collection.Seq;

public class RentalCalculator {

    public static Function1<Seq<Rental>, String> formatStatement =
            (rentals) -> rentals.foldLeft("", (statement, rental) -> statement + RentalCalculator.formatLine.apply(rental))
                    .concat(String.format("Total amount | %f", RentalCalculator.calculateRentalAmount.apply(rentals)));

    public static Function1<Seq<Rental>, Double> calculateRentalAmount =
            (rentals) -> {
                checkRentals(rentals);
                return rentals.map(Rental::getAmount).sum().doubleValue();
            };

    private static Function1<Rental, String> formatLine = rental ->
            String.format("%tF : %s | %f %n",
                    rental.getDate(),
                    rental.getLabel(),
                    rental.getAmount());

    private static void checkRentals(Seq<Rental> rentals) {
        if (rentals.isEmpty()) {
            throw new IllegalStateException("No rentals !!!");
        }
    }
}

File: scala/src/main/scala/RentalCalculator.scala
package com.ythirion

import java.lang.System.lineSeparator

class RentalCalculator(val rentals: List[Rental]) {
  private var _amount = .0
  private var _calculated = false

  def amount: Double = _amount
  def calculated: Boolean = _calculated

  def calculateRental: String = {
    if (rentals.isEmpty) throw new IllegalStateException("No rentals !!!")
    val result = new StringBuilder

    for (rental <- rentals) {
      if (!_calculated) this._amount += rental.amount
      result.append(formatLine(rental, _amount))
    }
    result.append(f"Total amount | ${this._amount}%.2f")
    _calculated = true

    result.toString
  }

  private def formatLine(rental: Rental, amount: Double) =
    f"${rental.date} : ${rental.label} | ${rental.amount}%.2f${lineSeparator()}"
}
